### ISSUE TITLE
[JENKINS-14656] Jenkins cannot guarantee that it will push the exact same build through a build pipeline if some of the steps are triggered manually

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/build-pipeline-plugin#196]
<p>Configure a build pipeline of two jobs, where the second job is triggered manually. The Copy Artifacts plugin in the second job cannot select the artifacts from the first job because TriggeredBuildSelector only checks for "hudson.model.Cause$UpstreamCause", and not for "au.com.centrumsystems.hudson.plugin.buildpipeline.BuildPipelineView$MyUserIdCause".</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=chrisr">chrisr</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-14656" target="_blank">Jenkins cannot guarantee that it will push the exact same build through a build pipeline if some of the steps are triggered manually</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/mindless.png" title="mindless's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=mindless">mindless</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Major
<li><b>component(s)</b>: build-pipeline-plugin, copyartifact-plugin
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2014-06-23T23:49:59+00:00
<li><b>votes</b>: 9
<li><b>watchers</b>: 13
<li><b>imported</b>: 20260107-091649
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>Configure a build pipeline of two jobs, where the second job is triggered manually. The Copy Artifacts plugin in the second job cannot select the artifacts from the first job because TriggeredBuildSelector only checks for "hudson.model.Cause$UpstreamCause", and not for "au.com.centrumsystems.hudson.plugin.buildpipeline.BuildPipelineView$MyUserIdCause".</p></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>Windows 7</code></li></ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>chrisr:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-14656?focusedId=166295&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-166295">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>We are trying to use Jenkins' build pipelines for continual deployment, which means that our pipelines steps would be:</p>

<p>a) Build
b) Deploy to Test Platform
c) Run systems tests
d) Deploy to Production Platform
e) Run smoke tests</p>

<p>So step (d) will always be triggered manually, and we MUST ensure that the artifacts we deploy from step (a) have also passed step (c).</p>
</pre>
</details>

</ul>
</details>

<p>We are trying to use Jenkins' build pipelines for continual deployment, which means that our pipelines steps would be:</p>

<p>a) Build<br/>
b) Deploy to Test Platform<br/>
c) Run systems tests<br/>
d) Deploy to Production Platform<br/>
e) Run smoke tests</p>

<p>So step (d) will always be triggered manually, and we MUST ensure that the artifacts we deploy from step (a) have also passed step (c).</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>masto:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-14656?focusedId=166604&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-166604">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I've been struggling with the same problem for the past couple of days.  It seems like both plugins are slightly broken; the Build Pipeline doesn't set up the environment exactly the same for its manually-triggered jobs as an auto-triggered downstream build does: for example, the log doesn't always say <tt>Started by upstream project "test_upstream" build number 30</tt> (although sometimes it does, and I haven't figured out the difference).  Setting a Run Parameter works consistently but only in one of the two pipelines I created.  I think Build Pipelines is a bit buggy but I don't understand enough about the internals to make sense of the symptoms.  Looking at build.xml from a manually triggered vs automatically triggered build, the manual one seems to have an amalgam of system configuration, the upstream project's configuration, and the downstream one.  It looks wrong to me.</p>

<p>Regardless, since the downstream job does always have upstream build information available, Copy Artifact ought to be able to use it.  It would be nice to have the option to use a standard Run Parameter rather than the special copy parameter that other plugins don't know how to set.</p>
</pre>
</details>

</ul>
</details>

<p>I've been struggling with the same problem for the past couple of days.  It seems like both plugins are slightly broken; the Build Pipeline doesn't set up the environment exactly the same for its manually-triggered jobs as an auto-triggered downstream build does: for example, the log doesn't always say <tt>Started by upstream project "test_upstream" build number 30</tt> (although sometimes it does, and I haven't figured out the difference).  Setting a Run Parameter works consistently but only in one of the two pipelines I created.  I think Build Pipelines is a bit buggy but I don't understand enough about the internals to make sense of the symptoms.  Looking at build.xml from a manually triggered vs automatically triggered build, the manual one seems to have an amalgam of system configuration, the upstream project's configuration, and the downstream one.  It looks wrong to me.</p>

<p>Regardless, since the downstream job does always have upstream build information available, Copy Artifact ought to be able to use it.  It would be nice to have the option to use a standard Run Parameter rather than the special copy parameter that other plugins don't know how to set.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>masto:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-14656?focusedId=166670&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-166670">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I did some more poking around to confirm the theory that Build Pipeline is adding multiple CauseActions to the build instead of just one CauseAction containing multiple causes.  To do this, I installed the Groovy plugin and added an "Execute system Groovy script" action to my downstream build:</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">def build = <span class="code-object">Thread</span>.currentThread().executable;
def actions = build.getActions(hudson.model.CauseAction);

println <span class="code-quote">"\nbuild.getActions(hudson.model.CauseAction) returned "</span> + actions.size() + <span class="code-quote">" action(s)"</span>;

actions.each { action -&gt;
  action.causes.each { cause -&gt;
    println <span class="code-quote">"("</span> + cause.<span class="code-keyword">class.</span>name + <span class="code-quote">") "</span> + cause.shortDescription;
  }
}

println <span class="code-quote">"\nbuild.getCause(hudson.model.Cause.UpstreamCause) returns:"</span>;

def cause = build.getCause(hudson.model.Cause.UpstreamCause);
println <span class="code-quote">"("</span> + cause?.class?.name + <span class="code-quote">") "</span> + cause?.shortDescription + <span class="code-quote">"\n"</span>;
</pre>
</div></div>

<p>When manually triggering a downstream job, this is what I get in the console output:</p>

<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>[EnvInject] - Loading node environment variables.
Building on master in workspace /apps/jenkins/home/jobs/cmasto_test_manual/workspace

build.getActions(hudson.model.CauseAction) returned 2 action(s)
(au.com.centrumsystems.hudson.plugin.buildpipeline.BuildPipelineView$MyUserIdCause) Started by user cmasto
(hudson.model.Cause$UpstreamCause) Started by upstream project "cmasto_test_build" build number 28

build.getCause(hudson.model.Cause.UpstreamCause) returns:
(null) null

Unable to find a build for artifact copy from: cmasto_test_build
Build step 'Copy artifacts from another project' marked build as failure
Finished: FAILURE
</pre>
</div></div>

<p>Note I have added the Copy Artifact plugin as well, since making that work is the primary reason I need to get the upstream cause working correctly.</p>

<p>It doesn't even fail consistently.. if I trigger the same build a second time, it usually ends up swapping the order of the CauseActions, and the first one wins:</p>

<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>Started by upstream project "cmasto_test_build" build number 28
[EnvInject] - Loading node environment variables.
Building on master in workspace /apps/jenkins/home/jobs/cmasto_test_manual/workspace

build.getActions(hudson.model.CauseAction) returned 2 action(s)
(hudson.model.Cause$UpstreamCause) Started by upstream project "cmasto_test_build" build number 28
(au.com.centrumsystems.hudson.plugin.buildpipeline.BuildPipelineView$MyUserIdCause) Started by user cmasto

build.getCause(hudson.model.Cause.UpstreamCause) returns:
(hudson.model.Cause$UpstreamCause) Started by upstream project "cmasto_test_build" build number 28

Copied 1 artifact from "Masto Build" build number 28
Finished: SUCCESS
</pre>
</div></div>
</pre>
</details>

</ul>
</details>

<p>I did some more poking around to confirm the theory that Build Pipeline is adding multiple CauseActions to the build instead of just one CauseAction containing multiple causes.  To do this, I installed the Groovy plugin and added an "Execute system Groovy script" action to my downstream build:</p>


<pre>
def build = <span class="code-object">Thread</span>.currentThread().executable;
def actions = build.getActions(hudson.model.CauseAction);

println <span class="code-quote">"\nbuild.getActions(hudson.model.CauseAction) returned "</span> + actions.size() + <span class="code-quote">" action(s)"</span>;

actions.each { action ->
  action.causes.each { cause ->
    println <span class="code-quote">"("</span> + cause.<span class="code-keyword">class.</span>name + <span class="code-quote">") "</span> + cause.shortDescription;
  }
}

println <span class="code-quote">"\nbuild.getCause(hudson.model.Cause.UpstreamCause) returns:"</span>;

def cause = build.getCause(hudson.model.Cause.UpstreamCause);
println <span class="code-quote">"("</span> + cause?.class?.name + <span class="code-quote">") "</span> + cause?.shortDescription + <span class="code-quote">"\n"</span>;
</pre>

<p>When manually triggering a downstream job, this is what I get in the console output:</p>



```
[EnvInject] - Loading node environment variables.
Building on master in workspace /apps/jenkins/home/jobs/cmasto_test_manual/workspace

build.getActions(hudson.model.CauseAction) returned 2 action(s)
(au.com.centrumsystems.hudson.plugin.buildpipeline.BuildPipelineView$MyUserIdCause) Started by user cmasto
(hudson.model.Cause$UpstreamCause) Started by upstream project "cmasto_test_build" build number 28

build.getCause(hudson.model.Cause.UpstreamCause) returns:
(null) null

Unable to find a build for artifact copy from: cmasto_test_build
Build step 'Copy artifacts from another project' marked build as failure
Finished: FAILURE

```

<p>Note I have added the Copy Artifact plugin as well, since making that work is the primary reason I need to get the upstream cause working correctly.</p>

<p>It doesn't even fail consistently.. if I trigger the same build a second time, it usually ends up swapping the order of the CauseActions, and the first one wins:</p>



```
Started by upstream project "cmasto_test_build" build number 28
[EnvInject] - Loading node environment variables.
Building on master in workspace /apps/jenkins/home/jobs/cmasto_test_manual/workspace

build.getActions(hudson.model.CauseAction) returned 2 action(s)
(hudson.model.Cause$UpstreamCause) Started by upstream project "cmasto_test_build" build number 28
(au.com.centrumsystems.hudson.plugin.buildpipeline.BuildPipelineView$MyUserIdCause) Started by user cmasto

build.getCause(hudson.model.Cause.UpstreamCause) returns:
(hudson.model.Cause$UpstreamCause) Started by upstream project "cmasto_test_build" build number 28

Copied 1 artifact from "Masto Build" build number 28
Finished: SUCCESS

```

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>derjohn:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-14656?focusedId=175940&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-175940">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Heyho,
same problem here <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/> After reading Christopher's excellent analysis I create a groovy script that injects the variable for me. It simply loops over the causes until it find the cause with class name hudson.model.Cause$UpstreamCause and then injects a UPSTREAM_BUILD_NUMBER env variable. This can be used in the copyartifact step to copy "specific build number". And yes, you need additionally the groovy plugin and envinject.</p>

<p>=== 8&lt; ===
import hudson.model.*
def upstreamBuildNumber
def build = Thread.currentThread().executable
def actions = build.getActions(hudson.model.CauseAction);</p>

<p>actions.each { action -&gt;
  action.causes.each { cause -&gt;
    if ( cause.class.name == 'hudson.model.Cause$UpstreamCause' ) {
      upstreamBuildNumber = "${cause.getUpstreamBuild()}"
    }
  }
}</p>



<p>def map = <span class="error">&#91;UPSTREAM_BUILD_NUMBER:upstreamBuildNumber&#93;</span>
return map </p>

<p>=== 8&lt; ===</p>


<p>HTH,
Andreas</p>
</pre>
</details>

</ul>
</details>

<p>Heyho,<br/>
same problem here <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/> After reading Christopher's excellent analysis I create a groovy script that injects the variable for me. It simply loops over the causes until it find the cause with class name hudson.model.Cause$UpstreamCause and then injects a UPSTREAM_BUILD_NUMBER env variable. This can be used in the copyartifact step to copy "specific build number". And yes, you need additionally the groovy plugin and envinject.</p>

<p>=== 8< ===<br/>
import hudson.model.*<br/>
def upstreamBuildNumber<br/>
def build = Thread.currentThread().executable<br/>
def actions = build.getActions(hudson.model.CauseAction);</p>

<p>actions.each { action -><br/>
  action.causes.each { cause -><br/>
    if ( cause.class.name == 'hudson.model.Cause$UpstreamCause' ) {<br/>
      upstreamBuildNumber = "${cause.getUpstreamBuild()}"<br/>
    }<br/>
  }<br/>
}</p>



<p>def map = <span class="error">&#91;UPSTREAM_BUILD_NUMBER:upstreamBuildNumber&#93;</span><br/>
return map </p>

<p>=== 8< ===</p>


<p>HTH,<br/>
Andreas</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>stormwind:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-14656?focusedId=176566&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-176566">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Good investigation and workaround so far!</p>

<p>I have a more complex situation where a parallel job uses the Join plugin to run multiple jobs in parallel then combines the results from multiple jobs.   However, the upstream project and thus UPSTREAM_BUILD_NUMBER is always just the last job to complete from the parallel set so it can't access the build numbers for the other parallel jobs.</p>
</pre>
</details>

</ul>
</details>

<p>Good investigation and workaround so far!</p>

<p>I have a more complex situation where a parallel job uses the Join plugin to run multiple jobs in parallel then combines the results from multiple jobs.   However, the upstream project and thus UPSTREAM_BUILD_NUMBER is always just the last job to complete from the parallel set so it can't access the build numbers for the other parallel jobs.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-14656?focusedId=184915&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-184915">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Pull #8 should solve this. Verified with Copy Artifact plugin’s <em>Upstream build that triggered this job</em>.</p>
</pre>
</details>

</ul>
</details>

<p>Pull #8 should solve this. Verified with Copy Artifact plugin’s <em>Upstream build that triggered this job</em>.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>jenkinstm:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-14656?focusedId=193130&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-193130">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Any idea when this fix will make it into a release?
Thanks</p>
</pre>
</details>

</ul>
</details>

<p>Any idea when this fix will make it into a release?<br/>
Thanks</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>ikedam:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-14656?focusedId=204459&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-204459">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Looks fixed in <a href="https://github.com/jenkinsci/build-pipeline-plugin/pull/8" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/build-pipeline-plugin/pull/8</a> , build-pipeline-1.4.</p>
</pre>
</details>

</ul>
</details>

<p>Looks fixed in <a href="https://github.com/jenkinsci/build-pipeline-plugin/pull/8" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/build-pipeline-plugin/pull/8</a> , build-pipeline-1.4.</p>
